src/main/java/carpetextra/utils/BlockPlacer.java

```diff
@@ -2,117 +2,167 @@
-                    .with(DispenserBlock.TRIGGERED, Boolean.FALSE);
-        }
-        else if (block instanceof PistonBlock)
-        {
-            return block.getDefaultState()
-                    .with(FacingBlock.FACING, Direction.byId(code))
-                    .with(PistonBlock.EXTENDED, Boolean.FALSE);
-        }
-        else if (block instanceof StairsBlock)
+        } else if (block instanceof StairsBlock) {
+            state = block.getPlacementState(context)//worldIn, pos, facing, hitX, hitY, hitZ, meta, placer)
+                    .with(StairsBlock.FACING, facing)
+                    .with(StairsBlock.HALF, (hitX >= 16) ? BlockHalf.TOP : BlockHalf.BOTTOM);
+        } else if (block instanceof WallMountedBlock) {
+            //unsupported
+            return null;
+        } else if (block instanceof ChestBlock)
+        //-x +x
+        //+x east -x west +z south -z north
         {
-            return block.getPlacementState(context)//worldIn, pos, facing, hitX, hitY, hitZ, meta, placer)
-                    .with(StairsBlock.FACING, Direction.byId(code % 16))
-                    .with(StairsBlock.HALF, ( hitX >= 16)?BlockHalf.TOP : BlockHalf.BOTTOM);
+            if (facing == Direction.SOUTH) {
+                if (IsBlockAttachableChest(block, facing, pos.west(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.LEFT);
+                }
+                if (IsBlockAttachableChest(block, facing, pos.east(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.RIGHT);
+                }
+            } else if (facing == Direction.WEST)  //-z +z
+            {
+                if (IsBlockAttachableChest(block, facing, pos.north(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.LEFT);
+                }
+                if (IsBlockAttachableChest(block, facing, pos.south(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.RIGHT);
+                }
+            } else if (facing == Direction.NORTH) //+x -x
+            {
+                if (IsBlockAttachableChest(block, facing, pos.east(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.LEFT);
+                }
+                if (IsBlockAttachableChest(block, facing, pos.west(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.RIGHT);
+                }
+            } else if (facing == Direction.EAST) //+z -z
+            {
+                if (IsBlockAttachableChest(block, facing, pos.south(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.LEFT);
+                }
+                if (IsBlockAttachableChest(block, facing, pos.north(), world)) {
+                    return state.with(ChestBlock.CHEST_TYPE, ChestType.RIGHT);
+                }
+            }
+            return state.with(ChestBlock.CHEST_TYPE, ChestType.SINGLE);
```

Comment:
Is this the only way of doing this?

Reply:
I guess it can use .Direction.rotateYClockwise() and rotateYCounterclockwise() instead of repeated facing checks, I'll try it